### ISSUE TITLE
Enable TLS automatically if a server offers it

### ISSUE THREAD

**Issue by Synchro** [PHPMailer/PHPMailer#387]
This was mentioned in #386, and just seems like a good idea. If we connect to a server (without using SSL) and it offers `STARTTLS` in its capability list, we should enable TLS by default, even if we have not specified `SMTPSecure = 'tls'`. I've got a working patch to do this and it works fine for me - I'm interested to hear what others think. Is it likely to run into a ton of SSL verification errors in PHP 5.6? Do we care?

**Comment by Synchro** [comment]
Pushed in [the autotls branch](https://github.com/PHPMailer/PHPMailer/tree/autotls) - please give it a try.

**Comment by cbergmann** [comment]
I think this is a very good idea. 
I also think that today it is no problem to get a cheap (free) valid SSL certificate but we are not in the control of the receiving smtp server (especially with directSend) this might not help us.
What do you think about a second new Flag "SMTPDontVerifyCerts" that would set the stream context up so it would not verify the certificates?

**Comment by Synchro** [comment]
We've already got that - the smtp connect function takes an options array that can set those in the stream parameters. It's not that easy to get at, but it's just a subclass away...

**Comment by jonataswalker** [comment]
This works on RoundCube

$config['smtp_server'] = 'tls://mail.example.net';
$config['smtp_conn_options'] = array(
   'ssl'         => array(
        'verify_peer'  => true,
        'verify_depth' => 3,
        'allow_self_signed' => true,
        'peer_name'  => 'mail.example.net',
        'cafile'  => '/home/admin/ssl/mail.example.net.pem'
   )
);

But this doesn't on PHPMailer
$options = array(
    'ssl' => array(
            'verify_peer'  => true,
            'verify_depth' => 3,
            'allow_self_signed' => true,
            'peer_name' => 'mail.example.net',
            'cafile' => '/home/admin/ssl/mail.example.net.pem'
    )
);

**Comment by Synchro** [comment]
Define "Doesn't work".

**Comment by jonataswalker** [comment]
I know this is a very repeated question but I've got RoundCube working with a self-signed ssl and php 5.6.8.

I got this PHPMailer debug:
Connection: opening to mail.mapeando.net:587, t=10, opt=array ()
Connection: opened
SERVER -> CLIENT: 220 mail.mapeando.net ESMTP Postfix
CLIENT -> SERVER: EHLO mapeando.net
SERVER -> CLIENT: 250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS250-ENHANCEDSTATUSCODES250-8BITMIME250 DSN
CLIENT -> SERVER: STARTTLS
SERVER -> CLIENT: 220 2.0.0 Ready to start TLS
CLIENT -> SERVER: QUIT

Connection: closed
SMTP connect() failed.
SMTP connect() failed.

**Comment by Synchro** [comment]
Try setting SMTPDebug to 3 or 4 for lower-level output.

**Comment by jonataswalker** [comment]
It is SMTPDebug = 3 already.
And SMTPDebug = 4

Connection: opening to mail.mapeando.net:587, t=10, opt=array ()
Connection: opened
SMTP -> get_lines(): $data was ""
SMTP -> get_lines(): $str is "220 mail.mapeando.net ESMTP Postfix"
SMTP -> get_lines(): $data is "220 mail.mapeando.net ESMTP Postfix"
SERVER -> CLIENT: 220 mail.mapeando.net ESMTP Postfix
CLIENT -> SERVER: EHLO mapeando.net
SMTP -> get_lines(): $data was ""
SMTP -> get_lines(): $str is "250-mail.mapeando.net"
SMTP -> get_lines(): $data is "250-mail.mapeando.net"
SMTP -> get_lines(): $data was "250-mail.mapeando.net"
SMTP -> get_lines(): $str is "250-PIPELINING"
SMTP -> get_lines(): $data is "250-mail.mapeando.net250-PIPELINING"
SMTP -> get_lines(): $data was "250-mail.mapeando.net250-PIPELINING"
SMTP -> get_lines(): $str is "250-SIZE 10240000"
SMTP -> get_lines(): $data is "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000"
SMTP -> get_lines(): $data was "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000"
SMTP -> get_lines(): $str is "250-VRFY"
SMTP -> get_lines(): $data is "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY"
SMTP -> get_lines(): $data was "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY"
SMTP -> get_lines(): $str is "250-ETRN"
SMTP -> get_lines(): $data is "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN"
SMTP -> get_lines(): $data was "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN"
SMTP -> get_lines(): $str is "250-STARTTLS"
SMTP -> get_lines(): $data is "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS"
SMTP -> get_lines(): $data was "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS"
SMTP -> get_lines(): $str is "250-ENHANCEDSTATUSCODES"
SMTP -> get_lines(): $data is "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS250-ENHANCEDSTATUSCODES"
SMTP -> get_lines(): $data was "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS250-ENHANCEDSTATUSCODES"
SMTP -> get_lines(): $str is "250-8BITMIME"
SMTP -> get_lines(): $data is "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS250-ENHANCEDSTATUSCODES250-8BITMIME"
SMTP -> get_lines(): $data was "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS250-ENHANCEDSTATUSCODES250-8BITMIME"
SMTP -> get_lines(): $str is "250 DSN"
SMTP -> get_lines(): $data is "250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS250-ENHANCEDSTATUSCODES250-8BITMIME250 DSN"
SERVER -> CLIENT: 250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS250-ENHANCEDSTATUSCODES250-8BITMIME250 DSN
CLIENT -> SERVER: STARTTLS
SMTP -> get_lines(): $data was ""
SMTP -> get_lines(): $str is "220 2.0.0 Ready to start TLS"
SMTP -> get_lines(): $data is "220 2.0.0 Ready to start TLS"
SERVER -> CLIENT: 220 2.0.0 Ready to start TLS
CLIENT -> SERVER: QUIT
SMTP -> get_lines(): $data was ""

Connection: closed
SMTP connect() failed.
SMTP connect() failed.

**Comment by jonataswalker** [comment]
I can confirm this is also working with Swift Mailer.

Using .... Swift_SmtpTransport::newInstance('mail.mapeando.net', 587, 'tls') and

        $options = array(
            'ssl' => array(
                    'verify_peer'  => true,
                    'verify_depth' => 3,
                    'allow_self_signed' => true,
                    'peer_name' => 'mail.mapeando.net',
                    'ciphers' => 'TLSv1+HIGH:!aNull:@STRENGTH',
                    'cafile' => '/home/admin/ssl/mail.mapeando.net.pem',
            )
        );

++ Starting Swift_SmtpTransport
<< 220 mail.mapeando.net ESMTP Postfix
>> EHLO mapeando.net
<< 250-mail.mapeando.net 250-PIPELINING 250-SIZE 10240000 250-VRFY 250-ETRN 250-STARTTLS 250-ENHANCEDSTATUSCODES 250-8BITMIME 250 DSN
>> STARTTLS
<< 220 2.0.0 Ready to start TLS
>> EHLO mapeando.net
<< 250-mail.mapeando.net 250-PIPELINING 250-SIZE 10240000 250-VRFY 250-ETRN 250-AUTH PLAIN LOGIN 250-AUTH=PLAIN LOGIN 250-ENHANCEDSTATUSCODES 250-8BITMIME 250 DSN
>> AUTH LOGIN
....
<< 235 2.7.0 Authentication successful
++ Swift_SmtpTransport started
>> MAIL FROM: <suporte@mapeando.net>
<< 250 2.1.0 Ok
>> RCPT TO: <jonataswalker@gmail.com>
<< 250 2.1.5 Ok
>> RCPT TO: <jonataswalker@yahoo.com>
<< 250 2.1.5 Ok
>> DATA
<< 354 End data with <CR><LF>.<CR><LF>
>> .
<< 250 2.0.0 Ok: queued as 7CBB4121163
++ Stopping Swift_SmtpTransport
>> QUIT
<< 221 2.0.0 Bye
++ Swift_SmtpTransport stopped

**Comment by Synchro** [comment]
Look at the first line of your output:
```
Connection: opening to mail.mapeando.net:587, t=10, opt=array ()
```

Options is empty. If you read above, you'll see there's a comment about that - access to the options array isn't straightforward in 5.2.9, however, I have just pushed a change in ab36b539a6a76cbf06b394fa64d8af60cb2b024c that exposes it in a simple way that will appear in the next release; you will just need to do this:

```php
$mail->SMTPOptions = array...
```

It's available now in HEAD, please give it a try.

**Comment by jonataswalker** [comment]
That's perfect man! Good job!

Connection: opening to mail.mapeando.net:587, timeout=300, options=array ( 'ssl' => array ( 'verify_peer' => true, 'verify_depth' => 3, 'allow_self_signed' => true, 'peer_name' => 'mail.mapeando.net', 'cafile' => '/home/admin/ssl/mail.mapeando.net.pem', ),)
Connection: opened
SERVER -> CLIENT: 220 mail.mapeando.net ESMTP Postfix
CLIENT -> SERVER: EHLO mapeando.net
SERVER -> CLIENT: 250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-STARTTLS250-ENHANCEDSTATUSCODES250-8BITMIME250 DSN
CLIENT -> SERVER: STARTTLS
SERVER -> CLIENT: 220 2.0.0 Ready to start TLS
CLIENT -> SERVER: EHLO mapeando.net
SERVER -> CLIENT: 250-mail.mapeando.net250-PIPELINING250-SIZE 10240000250-VRFY250-ETRN250-AUTH PLAIN LOGIN250-AUTH=PLAIN LOGIN250-ENHANCEDSTATUSCODES250-8BITMIME250 DSN
CLIENT -> SERVER: AUTH LOGIN
.....
SERVER -> CLIENT: 235 2.7.0 Authentication successful
CLIENT -> SERVER: MAIL FROM:<suporte@mapeando.net>
SERVER -> CLIENT: 250 2.1.0 Ok
CLIENT -> SERVER: RCPT TO:<jonataswalker@gmail.com>
SERVER -> CLIENT: 250 2.1.5 Ok
CLIENT -> SERVER: DATA
SERVER -> CLIENT: 354 End data with <CR><LF>.<CR><LF>
CLIENT -> SERVER: Date: Mon, 27 Apr 2015 16:32:46 -0300
CLIENT -> SERVER: To: jonataswalker@gmail.com
CLIENT -> SERVER: From: suporte@mapeando.net
CLIENT -> SERVER: Reply-To: suporte@mapeando.net
CLIENT -> SERVER: Subject: suporte@mapeando.net
CLIENT -> SERVER: Message-ID: <b7e2650a0be8880520ee7016093d7b8e@mapeando.net>
CLIENT -> SERVER: X-Priority: 3
CLIENT -> SERVER: X-Mailer: PHPMailer 5.2.9 (https://github.com/PHPMailer/PHPMailer/)
CLIENT -> SERVER: MIME-Version: 1.0
CLIENT -> SERVER: Content-Type: text/html; charset=UTF-8
CLIENT -> SERVER: Content-Transfer-Encoding: 8bit
CLIENT -> SERVER:
CLIENT -> SERVER: suporte@mapeando.net
CLIENT -> SERVER:
CLIENT -> SERVER: .
SERVER -> CLIENT: 250 2.0.0 Ok: queued as 6B5E3121151
CLIENT -> SERVER: QUIT
SERVER -> CLIENT: 221 2.0.0 Bye
Connection: closed

**Comment by Synchro** [comment]
autotls has now been merged into master and will be in the 5.2.10 release.
